File: 3862-find-the-smallest-balanced-index/3862-find-the-smallest-balanced-index.py
class Solution:
    def smallestBalancedIndex(self, nums: list[int]) -> int:
        if len(nums) == 0:
            return -1
        p = 1
        s = sum(nums)
        result = -1
        max_limit = 10**16

        for j in range(len(nums) - 1 , -1 , -1):
            s -= nums[j]
            if p == s:
                return j
            if p <= max_limit:
                p *= nums[j]
            else:
                p = max_limit + 1
        return result

File: 3862-find-the-smallest-balanced-index/3862-find-the-smallest-balanced-index.cpp
class Solution {
public:
    int smallestBalancedIndex(vector<int>& nums) {
        unsigned __int128 p = 1;
        long long s = 0;

        for(int i = 0; i < nums.size(); i++){
            s += nums[i];
        }

        for(int i = nums.size() - 1; i >= 0; i--){
            s -= nums[i];

            if(s == p){
                return i;
            }

            p *= nums[i];
            if(p >= s){
                break;
            }
        }
        return -1;
    }
};